freqtrade/data/dataprovider.py

```diff
@@ -150,6 +150,80 @@ def _add_external_df(
+        producer_name: str = "default"
+    ) -> Tuple[bool, int]:
+        """
+        Append a candle to the existing external dataframe
+
+        :param pair: pair to get the data for
+        :param timeframe: Timeframe to get data for
+        :param candle_type: Any of the enum CandleType (must match trading mode!)
+        :returns: False if the candle could not be appended, or the int number of missing candles.
+        """
+        pair_key = (pair, timeframe, candle_type)
+
+        if (producer_name not in self.__producer_pairs_df) \
+           or (pair_key not in self.__producer_pairs_df[producer_name]):
+            # We don't have data from this producer yet,
+            # or we don't have data for this pair_key
+            # return False and 1000 for the full df
+            return (False, 1000)
+
+        existing_df, _ = self.__producer_pairs_df[producer_name][pair_key]
+
+        # CHECK FOR MISSING CANDLES
+        timeframe_delta = to_timedelta(timeframe)  # Convert the timeframe to a timedelta for pandas
+        local_last = existing_df.iloc[-1]['date']  # We want the last date from our copy of data
+        incoming_first = dataframe.iloc[0]['date']  # We want the first date from the incoming data
+
+        # We have received this candle before, update our copy
+        # and return True, 0
+        if local_last == incoming_first:
+            existing_df.iloc[-1] = dataframe.iloc[0]
+            existing_data = (existing_df.reset_index(drop=True), _)
+
+            self.__producer_pairs_df[producer_name][pair_key] = existing_data
+            return (True, 0)
+
+        candle_difference = (incoming_first - local_last) / timeframe_delta
+
+        # If the difference divided by the timeframe is 1, then this
+        # is the candle we want and the incoming data isn't missing any.
+        # If the candle_difference is more than 1, that means
+        # we missed some candles between our data and the incoming
+        # so return False and candle_difference.
+        if candle_difference > 1:
+            return (False, candle_difference)
+
+        appended_df = self._append_candle_to_dataframe(existing_df, dataframe)
+
+        # Everything is good, we appended
+        self.__producer_pairs_df[producer_name][pair_key] = appended_df, last_analyzed
+        return (True, 0)
+
+    def _append_candle_to_dataframe(self, existing: DataFrame, new: DataFrame) -> DataFrame:
+        """
+        Append the `new` dataframe to the `existing` dataframe
+
+        :param existing: The full dataframe you want appended to
+        :param new: The new dataframe containing the data you want appended
+        :returns: The dataframe with the new data in it
+        """
+        if existing.iloc[-1]['date'] != new.iloc[-1]['date']:
+            existing = concat([existing, new])
+
+        # Only keep the last 1500 candles in memory
+        existing = existing[-1500:] if len(existing) > 1500 else existing
```

Comment:
You must reindex after this - otherwise the index will start floating from 1-1501, 2-1502, ... - as it's not reset again.